# Análisis Estratégico de Precios Mayoristas para el Sector Retail de Frutas y Hortalizas

**Autor: Ricardo Urdaneta**

Fecha: 24 julio 2025


---


**FRESCORP S.A.** es una cadena de supermercados FICTICIA, con fuerte presencia en la venta de abarrotes, carnes y productos envasados. Recientemente, han notado un creciente interés de sus clientes por productos frescos, saludables y de origen local.

Para capitalizar esta tendencia y competir con las grandes cadenas y ferias locales, la gerencia ha decidido lanzar una nueva línea de negocio: **"FRESCORP Verde"**, dedicada exclusivamente a la venta de frutas y hortalizas de alta calidad.

>La empresa no tiene experiencia en la compra a gran escala de productos perecederos. El mercado de frutas y hortalizas es volátil y los precios cambian constantemente. Antes de invertir en logística, refrigeración y personal, necesitan un **plan de negocio basado en datos** para asegurar la rentabilidad del proyecto y ofrecer precios competitivos a sus clientes.

---

### - Objetivos del Análisis:

Este análisis tiene como objetivo proporcionar una estrategia inicial para el lanzamiento de "FRESCORP Verde", respondiendo a preguntas críticas de negocio en tres áreas fundamentales.

---

### - Preguntas Clave:

#### Selección de Cartera de Productos
* ¿Qué productos tienen los precios de compra más estables a lo largo del año?
* ¿Qué productos tienen el menor costo de adquisición promedio?
* ¿Existen "canastas básicas" de productos (ej. tomate, cebolla, papa) cuyos precios combinados sean predecibles?

#### Estrategia de Precios y Promociones
* ¿Cuáles son los meses de "temporada baja" (precios más baratos) para productos populares como paltas, frutillas o uvas?
* ¿Qué productos sufren las mayores alzas de precio?
* ¿Podemos identificar productos "gancho" (baratos todo el año) y productos "premium" (volátiles pero con buen margen en temporada)?

#### Optimización de la Cadena de Suministro
* ¿Hay días específicos de la semana o quincenas del mes en los que los precios mayoristas tienden a ser más bajos?
* ¿Cómo fluctúa el precio de productos clave que se pueden almacenar por más tiempo (ej. papa, zapallo)?
---

#### Preparamos el ambiente

Importamos las librerias que usaremos para el analisis

In [1]:
# --- Librerías para Manipulación y Análisis de Datos ---
import pandas as pd  # Para manipulación de datos en estructuras tabulares
import numpy as np   # Para operaciones numéricas avanzadas
import datetime as dt  # Para manejar fechas y tiempos

# --- Librerías para Visualización de Datos ---
import matplotlib.pyplot as plt  # Para crear gráficos y visualizaciones
import seaborn as sns            # Para visualizaciones estadísticas más atractivas

# Asegura que los gráficos se muestren directamente en el notebook
%matplotlib inline

In [8]:
# Define la ruta del archivo CSV
file_path = 'data/lista_mayorista.csv'

# Carga el archivo CSV en un DataFrame de pandas
df = pd.read_csv(file_path)

#### Exploración de datos inicial
Exploramos la base de datos usando las funciones `df.head()` y `df.tail()`

In [11]:
# Muestra las primeras 5 filas del DataFrame
df.head()

Fecha  ID region                Region  \
0  2025-01-02          5  Región de Valparaíso   
1  2025-01-02          8     Región del Biobío   
2  2025-01-02         16       Región de Ñuble   
3  2025-01-02         16       Región de Ñuble   
4  2025-01-02          4    Región de Coquimbo   

                                Mercado                Subsector Producto  \
0                  Femacal de La Calera  Hortalizas y tubérculos   Acelga   
1            Vega Monumental Concepción  Hortalizas y tubérculos   Acelga   
2  Terminal Hortofrutícola Agro Chillán  Hortalizas y tubérculos   Acelga   
3  Terminal Hortofrutícola Agro Chillán  Hortalizas y tubérculos   Acelga   
4      Terminal La Palmera de La Serena  Hortalizas y tubérculos      Ají   

   Variedad / Tipo  Calidad    Unidad de comercializacion  \
0  Sin especificar  Primera  $/docena de atados (6 kilos)   
1  Sin especificar  Primera          $/atado 0,5 a 1 kilo   
2  Sin especificar  Primera          $/atado 0,5 a 1 kilo   
3  Sin especificar  Segunda          $/atado 0,5 a 1 kilo   
4          Inferno  Primera               $/caja 15 kilos   

                  Origen  Volumen Precio minimo Precio maximo  \
0  Provincia de Quillota      110     4000,0000     4000,0000   
1   Región Metropolitana      300      800,0000      800,0000   
2        Región de Ñuble      200      600,0000      600,0000   
3        Región de Ñuble      600      500,0000      500,0000   
4    Provincia de Limarí      400    15000,0000    17000,0000   

  Precio promedio ponderado  
0                 4000,0000  
1                  800,0000  
2                  600,0000  
3                  500,0000  
4                16000,0000

In [12]:
# Muestra las últimas 5 filas del DataFrame
df.tail()

Fecha  ID region                            Region  \
113506  2025-07-18         13  Región Metropolitana de Santiago   
113507  2025-07-18         13  Región Metropolitana de Santiago   
113508  2025-07-18         13  Región Metropolitana de Santiago   
113509  2025-07-18         13  Región Metropolitana de Santiago   
113510  2025-07-18         13  Región Metropolitana de Santiago   

                                          Mercado                Subsector  \
113506           Vega Central Mapocho de Santiago  Hortalizas y tubérculos   
113507  Mercado Mayorista Lo Valledor de Santiago  Hortalizas y tubérculos   
113508           Vega Central Mapocho de Santiago  Hortalizas y tubérculos   
113509  Mercado Mayorista Lo Valledor de Santiago  Hortalizas y tubérculos   
113510  Mercado Mayorista Lo Valledor de Santiago  Hortalizas y tubérculos   

        Producto  Variedad / Tipo  Calidad Unidad de comercializacion  \
113506  Cebollín  Sin especificar  Primera      $/paquete 36 unidades   
113507  Cebollín  Sin especificar  Primera      $/paquete 36 unidades   
113508  Cebollín  Sin especificar  Segunda      $/paquete 36 unidades   
113509  Cebollín  Sin especificar  Segunda      $/paquete 36 unidades   
113510  Cebollín  Sin especificar  Tercera      $/paquete 36 unidades   

                      Origen  Volumen Precio minimo Precio maximo  \
113506  Región Metropolitana      520     5000,0000     5200,0000   
113507  Región Metropolitana      900     5000,0000     5000,0000   
113508  Región Metropolitana      340     4000,0000     4500,0000   
113509  Región Metropolitana      600     4500,0000     4500,0000   
113510  Región Metropolitana      500     3800,0000     3800,0000   

       Precio promedio ponderado  
113506                 5100,0000  
113507                 5000,0000  
113508                 4250,0000  
113509                 4500,0000  
113510                 3800,0000

> Podemos identificar fácilmente que tenemos una mezcla de variables **categóricas, numéricas y temporales**.

Cada fila en nuestra base de datos representa un **registro de precio** en un momento específico. Es decir, nos muestra el **precio mayorista** para un **producto determinado** en una **fecha específica**.

Las columnas nos dan los detalles de este registro, incluyendo:

* **Variables Categóricas:** como `Producto` (ej: 'Tomate', 'Manzana Fuji') y `Categoría` (ej: 'Hortaliza', 'Fruta').
* **Variables Numéricas:** como el `Precio`.
* **Variable Temporal:** que es la `Fecha` del registro.

Esta estructura es ideal para analizar tendencias de precios, estacionalidad y volatilidad, que son los objetivos centrales del proyecto.

Ahora usamos las funciones `df.shape` y `df.info()` para obtener mas información de nuestra base de datos:

In [18]:
# Muestra la forma del DataFrame (número de filas y columnas)
df.shape

(113511, 14)

In [19]:
# Muestra información detallada del DataFrame, incluyendo tipos de datos y memoria utilizada
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113511 entries, 0 to 113510
Data columns (total 14 columns):
 #   Column                      Non-Null Count   Dtype 
---  ------                      --------------   ----- 
 0   Fecha                       113511 non-null  object
 1   ID region                   113511 non-null  int64 
 2   Region                      113511 non-null  object
 3   Mercado                     113511 non-null  object
 4   Subsector                   113511 non-null  object
 5   Producto                    113511 non-null  object
 6   Variedad / Tipo             113511 non-null  object
 7   Calidad                     113511 non-null  object
 8   Unidad de comercializacion  113511 non-null  object
 9   Origen                      113511 non-null  object
 10  Volumen                     113511 non-null  int64 
 11  Precio minimo               113511 non-null  object
 12  Precio maximo               113511 non-null  object
 13  Precio promedio ponderado   1

Con estas funciones obtenemos la primera radiografía de nuestra base de datos y nos revela:
1. La magnitud de la base de datos con 113511 registros en 14 columnas.
2. Los tipos de datos de las columnas y que debemos realizar ciertas conversiones para poder trabajar adecuadamente

#### EXPLICAR LAS CONVERSIONES QUE HAY QUE HACER:
datetime
las columnas de precios hubo que usar una funcion bucle para reemplazar la coma decimal por un punto para luego convertirlo a float


In [21]:
# Convertir la columna 'Fecha' a tipo datetime
df['Fecha'] = pd.to_datetime(df['Fecha'])

In [24]:
# Lista de las columnas de precios que convertiremos
columnas_precio = ['Precio minimo', 'Precio maximo', 'Precio promedio ponderado']

# Bucle para limpiar y convertir cada columna
for col in columnas_precio:
  # 1. Reemplaza la coma decimal por un punto
  df[col] = df[col].str.replace(',', '.', regex=False)
  # 2. Convierte la columna a tipo numérico (float)
  df[col] = pd.to_numeric(df[col])

# Verificamos que los tipos de datos han cambiado a float64
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113511 entries, 0 to 113510
Data columns (total 14 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Fecha                       113511 non-null  datetime64[ns]
 1   ID region                   113511 non-null  int64         
 2   Region                      113511 non-null  object        
 3   Mercado                     113511 non-null  object        
 4   Subsector                   113511 non-null  object        
 5   Producto                    113511 non-null  object        
 6   Variedad / Tipo             113511 non-null  object        
 7   Calidad                     113511 non-null  object        
 8   Unidad de comercializacion  113511 non-null  object        
 9   Origen                      113511 non-null  object        
 10  Volumen                     113511 non-null  int64         
 11  Precio minimo               113511 non-

#### Limpieza de datos:

Podemos apreciar rápidamente que no hay valores nulos, sin embargo y por buena práctica, lo confirmamos con la función `df.isnull().sum()`

In [ ]:
# Verificamos si hay valores nulos en el DataFrame
df.isnull().sum()

Fecha                         0
ID region                     0
Region                        0
Mercado                       0
Subsector                     0
Producto                      0
Variedad / Tipo               0
Calidad                       0
Unidad de comercializacion    0
Origen                        0
Volumen                       0
Precio minimo                 0
Precio maximo                 0
Precio promedio ponderado     0
dtype: int64

Así mismo también verificaremos los duplicados con la función `df.duplicated().sum()`

In [30]:
# Verificamos si hay duplicados en el DataFrame
df.duplicated().sum()

np.int64(0)

# Renombramos las columnas

In [35]:
# Renombramos, incluyendo un paso final para corregir dobles guiones bajos
df.columns = [col.lower().replace(' ', '_').replace('/', '_').replace('__', '_') for col in df.columns]

# Muestra los nombres corregidos para verificar
print(df.columns)

Index(['fecha', 'id_region', 'region', 'mercado', 'subsector', 'producto',
       'variedad_tipo', 'calidad', 'unidad_de_comercializacion', 'origen',
       'volumen', 'precio_minimo', 'precio_maximo',
       'precio_promedio_ponderado'],
      dtype='object')


In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113511 entries, 0 to 113510
Data columns (total 14 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   fecha                       113511 non-null  datetime64[ns]
 1   id_region                   113511 non-null  int64         
 2   region                      113511 non-null  object        
 3   mercado                     113511 non-null  object        
 4   subsector                   113511 non-null  object        
 5   producto                    113511 non-null  object        
 6   variedad_tipo               113511 non-null  object        
 7   calidad                     113511 non-null  object        
 8   unidad_de_comercializacion  113511 non-null  object        
 9   origen                      113511 non-null  object        
 10  volumen                     113511 non-null  int64         
 11  precio_minimo               113511 non-

---
# Trabajaremos en la estandarización de la columna unidad_de_comercializacion, explicar lo complicada que es

Paso 1: Extraer el Peso en KG (Casos Explícitos)
Primero, nos ocuparemos de los casos más sencillos, donde el peso en kilogramos está claramente indicado en la descripción.

Usaremos una expresión regular para buscar un número dentro de un paréntesis que esté asociado a "kg" o "kilo".

In [40]:
# Regex mejorado: busca un número seguido de un espacio y las palabras 'kilo', 'kilos' o 'kg'
regex_mejorado = r'(\d+)\s*(?:kilos|kilo|kg)'

# Paso 1: Extraer el peso usando el nuevo regex.
# Esto reemplazará la columna 'peso_kg' anterior con una mucho más completa.
df['peso_kg'] = df['unidad_de_comercializacion'].str.extract(regex_mejorado, expand=False)
df['peso_kg'] = pd.to_numeric(df['peso_kg'])

Paso 2: Calcular el Precio por Kilo
Con los pesos que acabamos de extraer, ya podemos calcular nuestra métrica estandarizada para una gran parte del dataset.

In [44]:
# Paso 2: (Re)calcular el precio por kilo con los nuevos datos
df['precio_por_kg'] = df['precio_promedio_ponderado'] / df['peso_kg']

# --- Verificación ---
print(f"Total de pesos extraídos con el código mejorado: {df['peso_kg'].notna().sum()}")

# Mostramos algunas filas para confirmar que funcionó correctamente
display(df[df['unidad_de_comercializacion'].str.contains('caja 10 kilos', na=False)].head())

Total de pesos extraídos con el código mejorado: 78873


fecha  id_region                            region  \
12 2025-01-02          9            Región de La Araucanía   
13 2025-01-02         13  Región Metropolitana de Santiago   
14 2025-01-02         16                   Región de Ñuble   
15 2025-01-02          8                 Región del Biobío   
16 2025-01-02          5              Región de Valparaíso   

                                      mercado                subsector  \
12                      Vega Modelo de Temuco  Hortalizas y tubérculos   
13  Mercado Mayorista Lo Valledor de Santiago  Hortalizas y tubérculos   
14       Terminal Hortofrutícola Agro Chillán  Hortalizas y tubérculos   
15                 Vega Monumental Concepción  Hortalizas y tubérculos   
16                       Femacal de La Calera  Hortalizas y tubérculos   

   producto variedad_tipo  calidad unidad_de_comercializacion  \
12      Ají       Inferno  Primera            $/caja 10 kilos   
13      Ajo         Chino  Primera            $/caja 10 kilos   
14      Ajo         Chino  Primera            $/caja 10 kilos   
15      Ajo         Chino  Primera            $/caja 10 kilos   
16      Ajo         Chino  Primera            $/caja 10 kilos   

                          origen  volumen  precio_minimo  precio_maximo  \
12  Región de Arica y Parinacota      125        25000.0        25000.0   
13                         China     1200        20000.0        20000.0   
14                         China       80        23000.0        23000.0   
15                         China      350        23000.0        24000.0   
16                         China       40        20000.0        20000.0   

    precio_promedio_ponderado  peso_kg  precio_por_kg  numero_unidades  
12                 25000.0000     10.0     2500.00000              NaN  
13                 20000.0000     10.0     2000.00000              NaN  
14                 23000.0000     10.0     2300.00000              NaN  
15                 23428.5714     10.0     2342.85714              NaN  
16                 20000.0000     10.0     2000.00000              NaN

## Estandarizar los productos que se venden por unidad.

El objetivo será crear dos nuevas columnas:

numero_unidades: Para guardar la cantidad de unidades (ej: 5, 36, etc.).

precio_por_unidad: Para calcular el precio de una sola unidad.

Paso 1: Extraer la Cantidad de Unidades
Primero, usaremos una expresión regular para extraer el número que aparece antes de la palabra "unidades".

In [43]:
# Regex: busca un número seguido de un espacio y la palabra 'unidades' o 'unidad'
regex_unidades = r'(\d+)\s*unidades?'

# Extraemos el número y lo guardamos en una nueva columna
df['numero_unidades'] = df['unidad_de_comercializacion'].str.extract(regex_unidades, expand=False)

# Convertimos a tipo numérico
df['numero_unidades'] = pd.to_numeric(df['numero_unidades'])

print("Valores de unidades extraídos:")
display(df[df['numero_unidades'].notna()][['unidad_de_comercializacion', 'numero_unidades']].head())

Valores de unidades extraídos:


unidad_de_comercializacion  numero_unidades
60         $/caja 50 unidades             50.0
61         $/caja 60 unidades             60.0
62         $/caja 40 unidades             40.0
63         $/caja 40 unidades             40.0
64         $/caja 40 unidades             40.0

Paso 2: Manejar Casos de "Unidad" Singular
Algunas descripciones podrían decir simplemente "unidad", lo que implica que es solo una. El código anterior no captura esto porque no hay un número explícito. Vamos a corregir esos casos y asignarles el valor 1.

In [45]:
# Buscamos filas que contienen 'unidad' pero donde no pudimos extraer un número
# y les asignamos el valor 1
condicion = df['unidad_de_comercializacion'].str.contains('unidad', na=False) & df['numero_unidades'].isnull()
df.loc[condicion, 'numero_unidades'] = 1

print("\nCasos de 'unidad' singular corregidos:")
display(df[condicion][['unidad_de_comercializacion', 'numero_unidades']].head())


Casos de 'unidad' singular corregidos:


unidad_de_comercializacion  numero_unidades
70                   $/unidad              1.0
77                   $/unidad              1.0
78                   $/unidad              1.0
79                   $/unidad              1.0
80                   $/unidad              1.0

Paso 3: Calcular el Precio por Unidad
Ahora que tenemos la cantidad de unidades, podemos calcular el precio estandarizado por unidad.

In [48]:
# Calculamos el precio por unidad
df['precio_por_unidad'] = df['precio_promedio_ponderado'] / df['numero_unidades']

# Mostramos un resumen final
print("\nResumen con la nueva columna 'precio_por_unidad':")
display(df[df['precio_por_unidad'].notna()][['producto', 'unidad_de_comercializacion', 'numero_unidades', 'precio_por_unidad']].head())


Resumen con la nueva columna 'precio_por_unidad':


producto unidad_de_comercializacion  numero_unidades  precio_por_unidad
60  Berenjena         $/caja 50 unidades             50.0         130.000000
61  Berenjena         $/caja 60 unidades             60.0          72.222222
62  Berenjena         $/caja 40 unidades             40.0         212.500000
63  Berenjena         $/caja 40 unidades             40.0         225.000000
64  Berenjena         $/caja 40 unidades             40.0         212.500000

Con esto, hemos estandarizado exitosamente los productos que se venden por unidad. Ahora tenemos  dos métricas muy poderosas para tu análisis: `precio_por_kg` y `precio_por_unidad`.

In [53]:
df.head(10)

fecha  id_region                            region  \
0 2025-01-02          5              Región de Valparaíso   
1 2025-01-02          8                 Región del Biobío   
2 2025-01-02         16                   Región de Ñuble   
3 2025-01-02         16                   Región de Ñuble   
4 2025-01-02          4                Región de Coquimbo   
5 2025-01-02          4                Región de Coquimbo   
6 2025-01-02         13  Región Metropolitana de Santiago   
7 2025-01-02          9            Región de La Araucanía   
8 2025-01-02         13  Región Metropolitana de Santiago   
9 2025-01-02          5              Región de Valparaíso   

                                     mercado                subsector  \
0                       Femacal de La Calera  Hortalizas y tubérculos   
1                 Vega Monumental Concepción  Hortalizas y tubérculos   
2       Terminal Hortofrutícola Agro Chillán  Hortalizas y tubérculos   
3       Terminal Hortofrutícola Agro Chillán  Hortalizas y tubérculos   
4           Terminal La Palmera de La Serena  Hortalizas y tubérculos   
5           Terminal La Palmera de La Serena  Hortalizas y tubérculos   
6  Mercado Mayorista Lo Valledor de Santiago  Hortalizas y tubérculos   
7                      Vega Modelo de Temuco  Hortalizas y tubérculos   
8  Mercado Mayorista Lo Valledor de Santiago  Hortalizas y tubérculos   
9                       Femacal de La Calera  Hortalizas y tubérculos   

  producto    variedad_tipo  calidad    unidad_de_comercializacion  \
0   Acelga  Sin especificar  Primera  $/docena de atados (6 kilos)   
1   Acelga  Sin especificar  Primera          $/atado 0,5 a 1 kilo   
2   Acelga  Sin especificar  Primera          $/atado 0,5 a 1 kilo   
3   Acelga  Sin especificar  Segunda          $/atado 0,5 a 1 kilo   
4      Ají          Inferno  Primera               $/caja 15 kilos   
5      Ají          Inferno  Segunda               $/caja 15 kilos   
6      Ají    Americana (o)  Primera               $/caja 25 kilos   
7      Ají    Americana (o)  Primera               $/caja 25 kilos   
8      Ají    Americana (o)  Segunda               $/caja 25 kilos   
9      Ají          Inferno  Primera               $/caja 15 kilos   

                  origen  volumen  precio_minimo  precio_maximo  \
0  Provincia de Quillota      110         4000.0         4000.0   
1   Región Metropolitana      300          800.0          800.0   
2        Región de Ñuble      200          600.0          600.0   
3        Región de Ñuble      600          500.0          500.0   
4    Provincia de Limarí      400        15000.0        17000.0   
5    Provincia de Limarí      240        12000.0        13000.0   
6    Provincia de Limarí       60        28000.0        28000.0   
7    Provincia de Limarí       55        35000.0        35000.0   
8    Provincia de Limarí       40        25000.0        25000.0   
9                Limache       45        29000.0        30000.0   

   precio_promedio_ponderado  peso_kg  precio_por_kg  numero_unidades  \
0                  4000.0000      6.0     666.666667              NaN   
1                   800.0000      1.0     800.000000              NaN   
2                   600.0000      1.0     600.000000              NaN   
3                   500.0000      1.0     500.000000              NaN   
4                 16000.0000     15.0    1066.666667              NaN   
5                 12500.0000     15.0     833.333333              NaN   
6                 28000.0000     25.0    1120.000000              NaN   
7                 35000.0000     25.0    1400.000000              NaN   
8                 25000.0000     25.0    1000.000000              NaN   
9                 29555.5555     15.0    1970.370367              NaN   

   precio_por_unidad  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
5                NaN  
6                NaN  
7                NaN  
8          

## Para saber cuántas filas no tienen valor ni en peso_kg ni en numero_unidades, usamos una condición que filtre los casos en los que ambas columnas son nulas.

Código para Contar las Filas sin Asignación
Este código te dará el número total de filas que no pudimos clasificar ni por kilo ni por unidad.

In [60]:
# Condición: busca filas donde 'peso_kg' es nulo Y 'numero_unidades' también es nulo
filas_sin_asignar = df[df['peso_kg'].isnull() & df['numero_unidades'].isnull()]

# Contamos cuántas filas cumplen esa condición
total_sin_asignar = len(filas_sin_asignar)

print(f"Número total de filas sin asignar: {total_sin_asignar}")

# Muestra las primeras 5 filas que no fueron asignadas
print("\nEjemplos de filas no asignadas:")
display(filas_sin_asignar[['unidad_de_comercializacion', 'producto']].head(10))

Número total de filas sin asignar: 5229

Ejemplos de filas no asignadas:


unidad_de_comercializacion  producto
23          $/docena de matas  Albahaca
24          $/docena de matas  Albahaca
25          $/docena de matas  Albahaca
26          $/docena de matas  Albahaca
27          $/docena de matas      Apio
28          $/docena de matas      Apio
29          $/docena de matas      Apio
30          $/docena de matas      Apio
31          $/docena de matas      Apio
32          $/docena de matas      Apio

Descubrimos que "docena" es una palabra clave que podemos convertir a un número específico.

Usaremos un código para buscar la palabra "docena" y asignar el valor 12 a la columna numero_unidades en las filas correspondientes.

Código para Asignar "Docena" = 12
Aplicaremos la misma lógica que usamos para la "unidad" singular.

In [61]:
# Condición: busca filas que contienen 'docena' y donde 'numero_unidades' aún no ha sido asignado.
condicion_docena = df['unidad_de_comercializacion'].str.contains('docena', na=False) & df['numero_unidades'].isnull()

# Asignamos el valor 12 a esas filas
df.loc[condicion_docena, 'numero_unidades'] = 12

# --- Verificación ---
print("Filas con 'docena' actualizadas a 12 unidades:")
display(df[condicion_docena][['unidad_de_comercializacion', 'numero_unidades']])

Filas con 'docena' actualizadas a 12 unidades:


unidad_de_comercializacion  numero_unidades
0       $/docena de atados (6 kilos)             12.0
23                 $/docena de matas             12.0
24                 $/docena de matas             12.0
25                 $/docena de matas             12.0
26                 $/docena de matas             12.0
...                              ...              ...
113281            $/docena de atados             12.0
113282            $/docena de atados             12.0
113286  $/docena de atados (2 kilos)             12.0
113358          $/docena de paquetes             12.0
113500          $/docena de paquetes             12.0

[5051 rows x 2 columns]

### Actualizar el Precio por Unidad
Ahora que hemos añadido más valores a numero_unidades, debemos volver a ejecutar el cálculo del precio_por_unidad para que estas nuevas filas también tengan su precio estandarizado.

In [62]:
# Recalculamos el precio por unidad para incluir los nuevos valores
df['precio_por_unidad'] = df['precio_promedio_ponderado'] / df['numero_unidades']

print("\nVerificación del precio por unidad para las docenas:")
display(df[condicion_docena][['unidad_de_comercializacion', 'precio_promedio_ponderado', 'numero_unidades', 'precio_por_unidad']])


Verificación del precio por unidad para las docenas:


unidad_de_comercializacion  precio_promedio_ponderado  \
0       $/docena de atados (6 kilos)                  4000.0000   
23                 $/docena de matas                  2900.0000   
24                 $/docena de matas                  3500.0000   
25                 $/docena de matas                  3277.7777   
26                 $/docena de matas                  2500.0000   
...                              ...                        ...   
113281            $/docena de atados                 11000.0000   
113282            $/docena de atados                 12000.0000   
113286  $/docena de atados (2 kilos)                  6000.0000   
113358          $/docena de paquetes                  9595.2380   
113500          $/docena de paquetes                  7000.0000   

        numero_unidades  precio_por_unidad  
0                  12.0         333.333333  
23                 12.0         241.666667  
24                 12.0         291.666667  
25                 12.0         273.148142  
26                 12.0         208.333333  
...                 ...                ...  
113281             12.0         916.666667  
113282             12.0        1000.000000  
113286             12.0         500.000000  
113358             12.0         799.603167  
113500             12.0         583.333333  

[5051 rows x 4 columns]

In [65]:
# Condición: busca filas donde 'peso_kg' es nulo Y 'numero_unidades' también es nulo
filas_sin_asignar = df[df['peso_kg'].isnull() & df['numero_unidades'].isnull()]

# Contamos cuántas filas cumplen esa condición
total_sin_asignar = len(filas_sin_asignar)

print(f"Número total de filas sin asignar: {total_sin_asignar}")

# Muestra las primeras 5 filas que no fueron asignadas
print("\nEjemplos de filas no asignadas:")
display(filas_sin_asignar[['unidad_de_comercializacion', 'producto']].head(10))

Número total de filas sin asignar: 1254

Ejemplos de filas no asignadas:


unidad_de_comercializacion      producto
156            $/caja 36 atados      Cilantro
157            $/caja 36 atados      Cilantro
158            $/caja 36 atados      Cilantro
847            $/caja 36 atados      Cilantro
848            $/caja 36 atados      Cilantro
891    $/atado 300 a 500 gramos      Espinaca
1282                     $/kilo  Poroto verde
1283                     $/kilo  Poroto verde
1459                  $/paquete      Albahaca
1460                  $/paquete      Albahaca